File: java/src/test/java/challenges/aoc2024/day3/service/ProgramServiceTest.java
package challenges.aoc2024.day3.service;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import challenges.aoc2024.day3.model.ComputerMemory;

class ProgramServiceTest {
    private ProgramService programService;

    @BeforeEach
    void setUp() {
        programService = new ProgramService();
    }

    @Test
    void shouldCalculateTotalOfMultiplicationInstructions() {
        // Given
        final ComputerMemory memory = ComputerMemory.of("xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))");

        // When
        final Integer total = programService.calculateTotalOfMultiplicationInstructions(memory);

        // Then
        assertEquals(161, total);
    }
}

File: scala/src/test/scala/challenges/aoc2024/day3/service/ProgramServiceTest.scala
package challenges.aoc2024.day3.service

import org.scalatest.BeforeAndAfter
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.prop.TableDrivenPropertyChecks.forAll
import org.scalatest.prop.TableFor2
import org.scalatest.prop.Tables.Table

class ProgramServiceTest extends AnyFunSuite with BeforeAndAfter {

  private var programService: ProgramService = _

  before {
    programService = ProgramService()
  }

  val cases: TableFor2[Boolean, Int] = Table(
    ("conditionalsIntact", "expectedTotal"),
    (false, 161),
    (true, 48)
  )
  test("should calculate total of multiplication instructions"):
    forAll(cases) { (conditionalsIntact: Boolean, expectedTotal: Int) =>
      // When
      val total: Int = programService.calculateTotalOfMultiplicationInstructions("xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))", conditionalsIntact)

      // Then
      assert(expectedTotal == total)
    }
}
